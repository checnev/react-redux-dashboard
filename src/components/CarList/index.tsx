import React, { useState, useMemo } from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableSortLabel,
  TableHead,
  TableRow,
  Paper,
  Link,
} from '@material-ui/core';
import { Link as LinkRoute } from 'react-router-dom';
import { Skeleton } from '@material-ui/lab';
import { STATUS_AS_STRING } from 'constants/car';
import type { CarReport } from 'reducers/vehicles/reports';
import type { CarWithReport } from 'selectors/cars';

const useStyles = makeStyles((theme: Theme) => createStyles({
  root: {}
}));

interface CarListProps {
  cars: CarWithReport[];
}

type SortKey = keyof CarReport;

export const compareReports = (
  a: CarWithReport,
  b: CarWithReport,
  sortKey: SortKey
): number => {
  if (a.report && b.report) {
    return a.report[sortKey] - b.report[sortKey];
  }
  
  return a.report? 1 : -1;
};

export const stableSort = (
  array: CarWithReport[],
  comparator: (a: CarWithReport, b: CarWithReport, sortKey: SortKey) => number,
  sortKey: SortKey,
  sortKind: 'asc' | 'desc',
) => {
  return array.slice().sort((a, b) => {
    if (sortKind === 'asc') {
      return comparator(a, b, sortKey);
    }
    return -comparator(a, b, sortKey);
  });
}

const sortFields: { name: SortKey, label: string }[] = [
  {
    name: 'countOfLeases',
    label: 'Кол-во Аренд',
  },
  {
    name: 'income',
    label: 'Доход',
  },
  {
    name: 'incidents',
    label: 'Происшествий',
  },
];

const CarList: React.FC<CarListProps> = ({ cars }: CarListProps) => {
  const classes  = useStyles();
  const [sortKey, setSortKey] = useState<SortKey | null>(null);
  const [sortKind, setSortKind] = useState<'asc' | 'desc'>('asc');
  const sortedCars = useMemo(() => {
    if (!sortKey) return cars;

    return stableSort(cars, compareReports, sortKey, sortKind)
  }, [cars, sortKey, sortKind]);

  const changeSort = (field: SortKey) => () => {
    if (sortKey === field) {
      setSortKind(sortKind === 'asc'? 'desc' : 'asc');
    } else {
      setSortKey(field);
    }
  };

  return (
    <TableContainer component={Paper}>
      <Table className={classes.root} aria-label="car list table">
        <TableHead>
          <TableRow>
            <TableCell>Гос. Номер</TableCell>
            <TableCell align="center">Марка, Модель</TableCell>
            {sortFields.map((field) => (
              <TableCell align="center" key={field.name} className="sortColumn">
                <TableSortLabel
                  onClick={changeSort(field.name)}
                  direction={sortKind}
                  active={sortKey === field.name}
                >
                  {field.label}
                </TableSortLabel>
              </TableCell>
            ))}
            <TableCell align="center">Статус</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {sortedCars && sortedCars.map((car) => (
            <TableRow key={car.id}>
              <TableCell component="th" scope="row">{car.stateNumber}</TableCell>
              <TableCell align="center">
                <Link component={LinkRoute} to={`/cars/${car.id}`} color="primary">{car.model}</Link>
              </TableCell>
              <TableCell align="center">
                {!car.report && <Skeleton />}
                {car.report && car.report.countOfLeases}
              </TableCell>
              <TableCell align="center">
                {!car.report && <Skeleton />}
                {car.report && car.report.income}
              </TableCell>
              <TableCell align="center">
                {!car.report && <Skeleton />}
                {car.report && car.report.incidents}
              </TableCell>
              <TableCell align="center">{STATUS_AS_STRING[car.status]}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default CarList;
