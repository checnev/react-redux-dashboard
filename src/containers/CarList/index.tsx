import { useSelector } from 'react-redux';
import { carWithReportsSelector } from 'selectors/cars';
import CarList from 'components/CarList';

const CarListContainer = () => {
  const carList = useSelector(carWithReportsSelector);
  return <CarList cars={carList} />;
}

export default CarListContainer;
