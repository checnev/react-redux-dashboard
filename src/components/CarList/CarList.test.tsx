import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import pretty from 'pretty';
import CarList from 'components/CarList';
import { BrowserRouter as Router } from 'react-router-dom';
import type { CarWithReport } from 'selectors/cars';

const carsWithReports: CarWithReport[] = [
  {
    id: 1,
    stateNumber: 'А 100 АА',
    model: 'Hyundai solaris 2014',
    status: 0,
    color: '#fff',
    imageUrl: '/images/solaris.jpg',
    coordinats: [1, 1],
    report: {
      income: 1000,
      countOfLeases: 4,
      incidents: 0,
    },
  },
  {
    id: 2,
    stateNumber: 'А 101 АА',
    model: 'Kia Rio 2014',
    status: 1,
    color: '#ccc',
    imageUrl: '/images/rio.jpg',
    coordinats: [1, 1],
    report: {
      income: 2000,
      countOfLeases: 2,
      incidents: 1,
    },
  },
];

let container: any = null;

beforeEach(() => {
  container = document.createElement("div");
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('CarList', () => {
  const setUp = (props: any) => (
    <Router>
      <CarList {...props} />
    </Router>
  );
  it('should render', () => {
    
    act(() => {
      render(setUp({ cars: carsWithReports }), container);
    });

    expect(
      pretty(container.innerHTML)
    ).toMatchSnapshot();
  });

  it('should render with sorted items', () => {

    act(() => {
      render(setUp({ cars: carsWithReports }), container);
    });

    const sortColumnButton = container.querySelector('th.sortColumn > span');

    act(() => {
      sortColumnButton.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });

    expect(
      pretty(container.innerHTML)
    ).toMatchSnapshot();
  });
});
