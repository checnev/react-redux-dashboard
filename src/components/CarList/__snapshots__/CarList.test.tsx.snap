// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarList should render 1`] = `
"<div class=\\"MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded\\">
  <table class=\\"MuiTable-root makeStyles-root-1\\" aria-label=\\"car list table\\">
    <thead class=\\"MuiTableHead-root\\">
      <tr class=\\"MuiTableRow-root MuiTableRow-head\\">
        <th class=\\"MuiTableCell-root MuiTableCell-head\\" scope=\\"col\\">Гос. Номер</th>
        <th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter\\" scope=\\"col\\">Марка, Модель</th>
        <th class=\\"MuiTableCell-root MuiTableCell-head sortColumn MuiTableCell-alignCenter\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Кол-во Аренд<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head sortColumn MuiTableCell-alignCenter\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Доход<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head sortColumn MuiTableCell-alignCenter\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Происшествий<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter\\" scope=\\"col\\">Статус</th>
      </tr>
    </thead>
    <tbody class=\\"MuiTableBody-root\\">
      <tr class=\\"MuiTableRow-root\\">
        <th class=\\"MuiTableCell-root MuiTableCell-body\\" role=\\"cell\\" scope=\\"row\\">А 100 АА</th>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary\\" href=\\"/cars/1\\">Hyundai solaris 2014</a></td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">4</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">1000</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">0</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">Свободен</td>
      </tr>
      <tr class=\\"MuiTableRow-root\\">
        <th class=\\"MuiTableCell-root MuiTableCell-body\\" role=\\"cell\\" scope=\\"row\\">А 101 АА</th>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary\\" href=\\"/cars/2\\">Kia Rio 2014</a></td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">2</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">2000</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">1</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">Занят</td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`CarList should render with sorted items 1`] = `
"<div class=\\"MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded\\">
  <table class=\\"MuiTable-root makeStyles-root-2\\" aria-label=\\"car list table\\">
    <thead class=\\"MuiTableHead-root\\">
      <tr class=\\"MuiTableRow-root MuiTableRow-head\\">
        <th class=\\"MuiTableCell-root MuiTableCell-head\\" scope=\\"col\\">Гос. Номер</th>
        <th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter\\" scope=\\"col\\">Марка, Модель</th>
        <th class=\\"MuiTableCell-root MuiTableCell-head sortColumn MuiTableCell-alignCenter\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-active\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Кол-во Аренд<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head sortColumn MuiTableCell-alignCenter\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Доход<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head sortColumn MuiTableCell-alignCenter\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Происшествий<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th>
        <th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter\\" scope=\\"col\\">Статус</th>
      </tr>
    </thead>
    <tbody class=\\"MuiTableBody-root\\">
      <tr class=\\"MuiTableRow-root\\">
        <th class=\\"MuiTableCell-root MuiTableCell-body\\" role=\\"cell\\" scope=\\"row\\">А 101 АА</th>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary\\" href=\\"/cars/2\\">Kia Rio 2014</a></td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">2</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">2000</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">1</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">Занят</td>
      </tr>
      <tr class=\\"MuiTableRow-root\\">
        <th class=\\"MuiTableCell-root MuiTableCell-body\\" role=\\"cell\\" scope=\\"row\\">А 100 АА</th>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary\\" href=\\"/cars/1\\">Hyundai solaris 2014</a></td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">4</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">1000</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">0</td>
        <td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter\\">Свободен</td>
      </tr>
    </tbody>
  </table>
</div>"
`;
